Support fast forward and rewind media actions when targeting Android 13 #972

defsub opened the pull request:
This PR makes fast forward and rewind media notification buttons work when targeting Android 13. It keeps the existing dart API and uses custom actions to add and handle events.

*If there's an open issue your PR is fixing, please list it here.*

fixes #972

## Pre-launch Checklist

- [x] I read the [CONTRIBUTING.md] and followed the process outlined there for submitting PRs.
- [x] My change is not breaking and lands in `minor` branch OR my change is breaking and lands in `major` branch.
- [x] If I'm the first to contribute to the next version, I incremented the version number in `pubspec.yaml` according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change (format: `* DESCRIPTION OF YOUR CHANGE (@your-git-username)`).
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I ran `dart analyze`.
- [x] I ran `dart format`.
- [x] I ran `flutter test` and all tests are passing.

<!-- Please consider also adding unit tests covering your new code. -->

<!-- Links -->
[CONTRIBUTING.md]: https://github.com/ryanheise/audio_service/blob/minor/CONTRIBUTING.md#making-a-pull-request
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

defsub committed:
fixes #972

EarningsCall commented:
We are interested in this change.  Is there any plan to merge it and accept the pull request?

ryanheise commented:
If this implementation works  via custom actions, I think I would want to consider this in the larger context of supporting any custom action in general. Perhaps it's not necessary for this PR to actually solve the problem of general custom actions, but we need to at least consider what such an API would look like now in order to ensure that this PR is future compatible with however that new API will turn out to work.

Any thoughts, @defsub , @EarningsCall ?

defsub commented:
I think what's needed is to add custom1, custom2, etc., to the MediaAction enum. The existing API should work fine since MediaControls can be added with the custom MediaAction (with custom android icon and label). On the Android side the the custrom control(s) would be added (similar to what I did with ff and rew)  and onCustomAction will report back the custom action name to be handled by dart code. Does this make sense to you? I think it's a fairly small change I can make if you agree on the approach.

ryanheise commented:
Actually I think that sounds quite elegant, although let me think about this.

According to:

https://developer.android.com/about/versions/13/behavior-changes-13

There can be up to 4 custom actions placed in the notification which are determined by an algorithm that fills in empty slots by custom actions if those slots are not filled by the actions they were originally intended for.

Now would adding 4 custom actions to the enum actually work? What if an app decides to use the custom enum items but not in the correct order? E.g. `custom1` and `custom4` but not `custom2` or `custom3`? I guess depending on your implementation, it wouldn't actually matter, and those unique enum values may help with the implementation. On the other hand, if those unique enum values don't help with the implementation, then you could instead just get by with a single `custom` enum value which an app could instantiate multiple times.

defsub commented:
With ff and rewind the custom actions I added are handled internally in java code. In the case of custom1, etc., the events would need to be handled in dart so I think the enum needs to be different to allow for different behavior. A single custom enum could be used but then we need to provide a custom action string for use in the callback. I had thought of invoking onCustomAction from java with "com.ryanheise.audioservice.CUSTOM_MEDIA_ACTION_1" which would map to the provided custom1 control/action. The string could be included with a new property in the MediaControl and use a single custom enum as well.

ryanheise commented:
It sounds like multiple enum values would be the way to go then.

There is still a question of whether the plugin should automatically map fast forward and rewind to custom actions on Android 13, or to just represent the native Android 13 behaviour more faithfully.

defsub commented:
I'll update the PR soon. My preference would be to automatically map them so the controls, when added, will work as before.

defsub commented:
I have the following working now. On Android 13 this adds 1 native and 4 custom controls: rewind(), fastForward(), and stop() are called w/o any extra dart code. The custom control displays a heart icon with a provided action string. Any number of custom controls can be added, internally they are of type `MediaAction.custom`.

```
controls: [
        MediaControl.rewind,
        if (_player.playing) MediaControl.pause else MediaControl.play,
        MediaControl.stop,
        MediaControl.fastForward,
        MediaControl.custom(androidIcon: "drawable/ic_baseline_favorite_24",
            label: "favorite", customAction: "favorite"),
      ],
```

```
  @override
  Future<dynamic> customAction(String name, [Map<String, dynamic>? extras]) {
    if (kDebugMode) {
      print(name);
    }
    return super.customAction(name, extras);
  }

```

Thus far I've yet to get custom controls working on my Android 9 phone. The issue is that the custom actions are added to the PlaybackState which is part of the MediaSession. The notification on Android 9 doesn't seem to be aware of these actions, unlike Android 13. I'm compiling and targeting 33 in all cases.

Finally, I also changed the platform interface to accommodate `custom` and the app provided action string.

```
--- a/audio_service_platform_interface/lib/audio_service_platform_interface.dart
+++ b/audio_service_platform_interface/lib/audio_service_platform_interface.dart
@@ -260,6 +260,9 @@ enum MediaActionMessage {
   seekBackward,
   seekForward,
   setSpeed,
+
+  ///
+  custom,
 }
 
 class MediaControlMessage {
@@ -273,17 +276,22 @@ class MediaControlMessage {
   /// The action to be executed by this control
   final MediaActionMessage action;
 
+  ///
+  final String? customAction;
+
   @literal
   const MediaControlMessage({
     required this.androidIcon,
     required this.label,
     required this.action,
+    this.customAction
   });
 
   Map<String, dynamic> toMap() => <String, dynamic>{
         'androidIcon': androidIcon,
         'label': label,
         'action': action.index,
+        if (customAction != null) 'customAction': customAction
       };
 }

```

Let me know if you agree with this general approach and I'll update the PR for further review.

defsub committed:
Merge branch 'minor' of github.com-defsub:defsub/audio_service into custom_actions_android13

ryanheise commented:
If you just used 4 distinct enum values `custom1`, ..., `custom4`, would that work on the current architecture without having to modify the platform interface?

defsub commented:
At a minimum the platform interface `MediaActionMessage` enum would need to support either `custom` or `custom1`, ..., `custom4`. And then the `customAction()` callback could be use as-is, passing the corresponding custom control # in the action name and/or extras. I think that's the smallest change. I was leaning toward allowing the caller to provide the callback name and using a single `custom` enum, but it is a bigger change.

ryanheise commented:
Ah yes, of course. So if we use multiple enum values `custom1`, `custom2`..., then we could potentially hard code the action string to `CUSTOM_1`, `CUSTOM_2`.... and do away with this extra `customAction` field. Although if it's an important use case to be able to choose the action string for a custom action, then there would be very little value in having multiple enum values when a single `custom` would suffice.

My gut feeling is that someone somewhere will want to be able to choose the action string.

Looking at the Android docs for `CustomAction`, it is also possible to pass in an extras bundle. So I wonder whether it would be worth introducing a new class to encapsulate this. Although I'm all for making minimal changes to the current API, and your idea elegantly does that, I also want to ensure that it is future compatible with whatever will need to happen in the future. So we are finding the optimal balance between minimising breaking changes now, and minimising breaking changes later.

@nt4f04uNd any thoughts yourself on the proposal, since I think you originally proposed the feature of custom actions?

nt4f04uNd commented:
Will take a look at this this weekend

tgrushka commented:
@defsub thanks for your fixes. I published this fork because we could not update Flutter without SDK 33 and this issue was blocking: https://pub.dev/packages/audio_service_13. At least the nudge back and forward buttons are now working, which is exactly what we needed.

nt4f04uNd reviewed (approved):
> Looking at the Android docs for `CustomAction`, it is also possible to pass in an extras bundle. So I wonder whether it would be worth introducing a new class to encapsulate this. Although I'm all for making minimal changes to the current API, and your idea elegantly does that, I also want to ensure that it is future compatible with whatever will need to happen in the future. So we are finding the optimal balance between minimising breaking changes now, and minimising breaking changes later.

I think if we ever introduce an API to pass extras to custom actions, this will not interfere in any way with changes in this PR. I think it's okay to merge this now as-is, since this fixes a broken functionality on Android 13, without changing the current API.

defsub committed:
Updates for custom action support in Android 13

Supports fast forward, rewind, and stop internally without any API changes.
Supports app defined custom actions with custom icons.

defsub commented:
I pushed my updated branch for review. (there isn't support for extras as of now) There's an example app as well so please take a look.

![Screenshot_20230201_132504_2](https://user-images.githubusercontent.com/18022414/218562425-eff2a641-eec0-4077-9b76-1f336c316913.png)

ryanheise commented:
Well I think if we are going to add an extras field later, we are going to run into the problem of deciding where it should go. So the current PR defines the addition of one new field:

```dart
public class MediaControl {
    public final String icon;
    public final String label;
    public final long actionCode;
    public final String customAction;
```

Now the first thing I would say is that this is already a breaking change since `customAction` is required. To make it backwards compatible, this should be nullable:

```dart
public class MediaControl {
    public final String icon;
    public final String label;
    public final long actionCode;
    public final String? customAction;
```

But the second thing is that if we are going to potentially add further fields associated with custom actions like this:

```dart
public class MediaControl {
    public final String icon;
    public final String label;
    public final long actionCode;
    public final String? customAction;
    public final Map<String, dynamic>? customActionExtras;
```

Then we have an inelegant API where two independent nullable fields actually should have dependent nullability. So for example, if `customAction` is null, then `customActionExtras` should also be null. I think if we are eventually going to add this, then a more elegant API design would be to encapsulate these linked nullables into a separate class and make the whole thing nullable to capture that dependency:

```dart
public class MediaControl {
    public final String icon;
    public final String label;
    public final long actionCode;
    public final CustomAction? customAction;
```

This would allow adding further fields to the encapsulated class over time without affecting this original design.

defsub commented:
@ryanheise I think you're looking at the java side of the code not dart. In the dart MediaControl, customAction is nullable as expected. I can go ahead and make some changes to support extras, however I do see limited value since similar things can be accomplished with varying action names.

ryanheise commented:
Ah, I was indeed looking at the Java code, the Dart has correct nullability.

I certainly agree that extras might have limited utility, although the main consideration for extras is with regards to future/backward compatibility. Although there are many features in the native framework that most people will not use, history tends to show that eventually someone will ask for it, and on that basis I have found it helpful to design the API more faithfully to the native API, or at least design it in a way that is future compatible with such a change.

defsub committed:
Update custom action support to use CustomAction

Support name and extras within CustomAction.
Updated example CustomAction favorite and custom icon for fast forward.
Updated platform interface version to 0.2.0.

defsub commented:
Please take a look at the updated PR with CustomAction support. The updated example uses action name and extras.

ryanheise commented:
```java
    PlaybackStateCompat.CustomAction createCustomAction(MediaControl action) {
        int iconId = getResourceId(action.icon);
        Bundle extras = new Bundle();
        extras.putLong(EXTRA_ACTION_CODE, action.actionCode);
        if (action.customAction != null) {
            extras.putString(EXTRA_CUSTOM_ACTION_NAME, action.customAction.name);
            if (action.customAction.extras != null) {
                extras.putBundle(EXTRA_CUSTOM_ACTION_EXTRAS, mapToBundle(action.customAction.extras));
            }
        }
        PlaybackStateCompat.CustomAction.Builder builder =
                new PlaybackStateCompat.CustomAction.Builder(
                        CUSTOM_MEDIA_BUTTON_ACTION, action.label, iconId).setExtras(extras);
        return builder.build();
    }
```

OK, so what I'm thinking here is that the reason people might want that extras field for custom actions is for interoperability purposes with other native Android code where they send a custom action over the transport controls -- and for those exact parameters sent from the Android side to be delivered to the Flutter side.

In the above code, it looks like that is not what's happening, instead the extras are being wrapped inside extras, using the `EXTRA_CUSTOM_ACTION_EXTRAS` key.

Maybe you have already thought about this and have reasons behind the nesting, but wouldn't it be possible to just pass the extras directly as the extras, rather than extras inside the extras?

defsub commented:
I'm not sure how that would work. The custom media control is added to the Android side and when used a custom intent (name + extras) is sent and eventually sent back to the flutter app via customAction. Each custom control needs a unique action name so the flutter side knows what control was used. Extras were added for extra stuff. That's why the name and extras are stored inside the intent. The extras provided by the flutter side come all the way back; see the new example app.

It seems you're thinking the custom media control would send an intent to another native receiver. Like say send an email or text or something. How would the flutter side know what happened? I wasn't thinking about enabling other things and instead simply enabling custom actions to work back and forth.

ryanheise commented:
I'm thinking more that the parent extras are the very extras that we are trying to expose equally on the flutter and the native side, so all of these extras really point to the same thing:

- the extras of a `CustomAction`
- the extras parameter of `onCustomAction` in Flutter
- the extras parameter of `onCustomAction` in native code

It's true we need an action name to identify the action, but all 3 of these have both an action name and extras which are supposed to have the same purpose and correspond to each other, and so rather than encode the action name and extras as a nested bundle within the extras, I am suggesting that the name and extras map directly to their equivalents in the native side.

So in the builder (in your code above), you call `setExtras` but there is also an action name in the first parameter of the builder's constructor, where you are currently passing in `CUSTOM_MEDIA_BUTTON_ACTION`. I am suggesting that you just pass the custom action name directly as that first parameter, since that is what it exactly corresponds to between the Flutter and native sides. Then in the following method, it sort of suggests that we shouldn't have an if/else and just pass everything through to be handled on the Flutter side:

```java
        public void onCustomAction(String action, Bundle extras) {
            if (listener == null) return;
            if (CUSTOM_MEDIA_BUTTON_ACTION.equals(action)) {
                long actionCode = extras.getLong(EXTRA_ACTION_CODE, -1);
                String actionName = extras.getString(EXTRA_CUSTOM_ACTION_NAME);
                if (actionName != null && actionName.length() > 0) {
                    Bundle actionExtras = extras.getBundle(EXTRA_CUSTOM_ACTION_EXTRAS);
                    if (actionExtras == null) {
                        actionExtras = Bundle.EMPTY;
                    }
                    listener.onCustomAction(actionName, actionExtras);
                }
                else if (actionCode == PlaybackStateCompat.ACTION_STOP) {
                    listener.onStop();
                } else if (actionCode == PlaybackStateCompat.ACTION_REWIND) {
                    listener.onRewind();
                } else if (actionCode == PlaybackStateCompat.ACTION_FAST_FORWARD) {
                    listener.onFastForward();
                } else {
                    listener.onCustomAction(action, extras);
                }
            } else {
                listener.onCustomAction(action, extras);
            }
        }
```

I see that this is where the logic is to automatically handle some of these actions, but I think we could perhaps just have a default implementation of `onCustomAction` on the Flutter side which does the same thing, but that way we also get to maintain an exact correspondence between the Flutter and native API.

defsub commented:
The java `onCustomAction` would basically be `listener.onCustomAction(action, extras)` with as you said, a default flutter `customAction` handling stop, rewind, fast forward and others as needed. I think that would work fine and it's a great suggestion. Were you suggesting earlier that this could enable additional custom native interaction? I'm not clear how that would happen w/o additional listeners.

ryanheise commented:
The native interaction part is basically referring to the fact that a native Android part of the app could use standard Android APIs to get a hold of the `TransportControls` object and invoke `sendCustomAction` directly by passing in the action name and extras, and here it would be nice if the caller did not have to encode the extras inside the extras, and could just pass the extras as the extras. Now if you use this native API, no additional listeners are required since this native API hooks into the standard listeners that we already have established.

defsub committed:
Changed custom action name and extras behavior

Incorporated PR feedback to make custom action really custom end to end.
Default implementation is provided and used in example app.
Example also includes a custom favorite action.

defsub commented:
@ryanheise please take a look at the updated PR.

ryanheise commented on audio_service/lib/audio_service.dart:
```diff
@@ -750,6 +753,72 @@ class _MediaItemCopyWith extends MediaItemCopyWith {
+class CustomAction {
+  /// Custom action name for stop.
+  static const stop = 'stop';
+  /// Custom action control for stop.
```
Do you use dart format? I think it would insert a blank line above this comment to separate it from the previous field.

ryanheise commented:
Thanks @defsub , I think the solution is looking a lot cleaner and simpler now. I'll continue reviewing and drop some comments on the code.

One thing regarding semantic versioning, I don't think this particular change to the platform interface is a breaking change, with regard to the proposed version number.

Presumably the functionality is working equally well for you as compared to your previous version?

ryanheise commented on audio_service/lib/audio_service.dart:
```diff
@@ -843,17 +912,43 @@ class MediaControl {
+      {required this.androidIcon,
+      required this.label,
+      required String name,
+      Map<String, dynamic>? extras})
```
Trailing comma missing (I should probably add this as a lint).

ryanheise commented on audio_service/lib/audio_service.dart:
```diff
@@ -843,17 +912,43 @@ class MediaControl {
+      {required this.androidIcon,
+      required this.label,
+      required this.action,
+      this.customAction});
```
Trailing comma

ryanheise commented on audio_service/lib/audio_service.dart:
```diff
@@ -843,17 +912,43 @@ class MediaControl {
+          {String? androidIcon,
+          String? label,
+          MediaAction? action,
+          CustomAction? customAction}) =>
```
Trailing comma

ryanheise commented on audio_service/lib/audio_service.dart:
```diff
@@ -843,17 +912,43 @@ class MediaControl {
+          androidIcon: androidIcon ?? this.androidIcon,
+          label: label ?? this.label,
+          action: action ?? this.action,
+          customAction: customAction ?? this.customAction);
```
Trailing comma

ryanheise commented on audio_service/lib/audio_service.dart:
```diff
@@ -843,17 +912,43 @@ class MediaControl {
         androidIcon: androidIcon,
         label: label,
         action: MediaActionMessage.values[action.index],
+        customAction: customAction?._toMessage(),
```
I approve of this comma ;-)

ryanheise commented on audio_service_platform_interface/lib/audio_service_platform_interface.dart:
```diff
@@ -273,20 +276,43 @@ class MediaControlMessage {
     required this.androidIcon,
     required this.label,
     required this.action,
+    this.customAction
```
Trailing comma

ryanheise commented on audio_service_platform_interface/lib/audio_service_platform_interface.dart:
```diff
@@ -273,20 +276,43 @@ class MediaControlMessage {
         'androidIcon': androidIcon,
         'label': label,
         'action': action.index,
+        if (customAction != null) 'customAction': customAction?.toMap()
```
Trailing comma

defsub commented:
Thanks for the feedback. I'll push the syntax changes shortly. Regarding versioning, my reading of semver.org says that going from 0.1.0 to 0.2.0, adds functionality in a backward compatible way. Using 0.1.1 would seemingly only be for backward compatible bug fixes.

defsub committed:
Trailing comma and syntax updates

ryanheise commented:
Thanks @defsub I will review in the morning.

Pub has a slightly different interpretation of semantic versioning when the version number is less than 1.0.0:

https://dart.dev/tools/pub/versioning#semantic-versions

The first non-zero digit acts as the major version and so increasing it acts as a breaking change. The pub tool uses this interpretation when resolving which versions will be compatible.

Edit: I may have simplified that a bit, if the first digit is zero, the second digit acts like a major version, but if the second digit is also zero the second digit still acts like a major version. I.e. the third digit is never a major version.

ryanheise commented on audio_service_platform_interface/pubspec.yaml:
```diff
@@ -3,7 +3,7 @@ description: A common platform interface for the audio_service plugin. Different
 # NOTE: We strongly prefer non-breaking changes, even at the expense of a
 # less-clean API. See https://flutter.dev/go/platform-interface-breaking-changes
-version: 0.1.0
+version: 0.2.0
```
Since it's a non-breaking change, 0.1.1 would be more indicative.

ryanheise commented on audio_service/lib/audio_service.dart:
```diff
@@ -3275,6 +3383,23 @@ mixin QueueHandler on BaseAudioHandler {
 
+/// This mixin provides default implementations for stop, fastForward, and
+/// rewind custom controls.
+mixin CustomActionHandler on BaseAudioHandler {
```
Since the native side is hard coded to treat these actions as custom actions on Android 13, and the Dart side is not hard coded and requires manually adding the mixin, it may make more sense to transfer this code into the default implementation of `customAction` in `BaseAudioHandler` similar to what happens in `click`.

defsub replied:
There's not much hard coded on the native side anymore. The only code that remains checks for Android 13 and forces those actions to be custom. However, they must already be defined as custom actions to work so this is unnecessary. I'll look at removing this code on the native side and then I'd think keeping the mixin as-is may make more sense.

ryanheise replied:
I'm referring to this part which activates even when `customAction == null`.

```java
    private boolean needCustomMediaControl(MediaControl control) {
        if (control.customAction != null) {
            return true;
        }

        // Android 13 changes MediaControl behavior as documented here:
        // https://developer.android.com/about/versions/13/behavior-changes-13
        // The below actions will be added to slots 1-3, if included.
        // 1 - ACTION_PLAY, ACTION_PLAY
        // 2 - ACTION_SKIP_TO_PREVIOUS
        // 3 - ACTION_SKIP_TO_NEXT
        // Custom actions will use slots 2-5 if included.
        // - ACTION_STOP
        // - ACTION_FAST_FORWARD
        // - ACTION_REWIND
        return (Build.VERSION.SDK_INT >= 33 &&
                (control.actionCode == PlaybackStateCompat.ACTION_STOP ||
                control.actionCode == PlaybackStateCompat.ACTION_FAST_FORWARD ||
                control.actionCode == PlaybackStateCompat.ACTION_REWIND));
    }
```

I think your original intention was to make these previously supported actions continue to work in Android 13 by default, so if a pre-13 app has `stop` in the `controls` parameter, this will continue to show in Android 13 by default. However, the button will do nothing unless you also add the mixin, so it seems it's sort of half doing the backwards compatibility thing (showing it but not routing it).

If you now think Android 13 shouldn't re-route these 3 actions by default, then it would make sense to also not show them by default, so I would either want to have these 3 actions a) all shown and routed by default, or b) not shown and not routed. I'm of two minds about it (and I guess you are too!)

defsub replied:
That's correct. I was going to come up with a more elegant approach but didn't get very far. Since I changed the build target to 33 (Android 13) that should mean that even pre-13 apps would need to use custom actions for `stop`, `fast forward`, and `rewind`. I could imagine  flutter code doing both - use `CustomAction.stopControl` and `MediaControl.stop` at the same time but that's ugly.

I'm now thinking it was cleaner before having  `stop`, `fast forward`, and `rewind` just work w/o doing any extra work. Any thoughts?

ryanheise commented:
By the way, I'm happy to make the above adjustments myself if you approve of my comment on mixins.

(Note I'm in the middle of moving apartments, probably my Internet should be back up by the end of the week.)

ryanheise commented:
I've got the internet reconnected at my new apartment, so I'd like to get this one merged today. Any objections if I fold the mixin into the base audio handler based on my analysis above?

defsub commented:
Good to hear you're (re)connected again. Thanks for the push on this one. I left a comment above. Let me know your thoughts.

defsub committed:
better support for existing mediacontrols

defsub commented:
@ryanheise I pushed some code that is backward compatible. If on Android 13 and `stop`, `fast forward`, or `rewind` controls are added, a custom action is automatically added to ensure the controls work. All other custom actions remain as-is.

defsub committed:
another android13 example, fixed pubspecs

defsub commented:
I'm pretty happy with it now. I added `example_android13.dart` that I've tested on phones with API level 28 and 33. `stop`, `fast forward`, and `rewind` code has the same controls yet works differently internally. It's also possible to do everything with custom actions as I have it in `example_custom_action.dart`. 

Restating what I changed, I automatically add the custom action handling when running on API 33 to be backward compatible. There's also custom actions defined in `audio_service.dart` and the corresponding `CustomActionHandler` mixin that can be used.

ryanheise commented on audio_service/android/src/main/java/com/ryanheise/audioservice/AudioService.java:
```diff
@@ -1082,7 +1068,15 @@ public void onSetShuffleMode(int shuffleMode) {
+            if (CUSTOM_ACTION_STOP.equals(action)) {
+                listener.onStop();
+            } else if (CUSTOM_ACTION_FAST_FORWARD.equals(action)) {
+                listener.onFastForward();
```
I'm unsure now whether want to handle these backwards compatible actions here in Java or over in Dart in the `CustomActionHandler` mixin. It looks like you're doing this in two places.

defsub replied:
The problem is that the Dart code doesn't easily know whether or not to use custom actions. On one hand, all new code can use custom actions and not worry about compatibility issues. However, existing code using deprecated actions, without being changed, will no longer work due to Android 13 changes. That's why I added back the java code to ensure existing code keeps working no matter the Android version. And, to be safe in the future, there are now custom actions to do whatever you want.

ryanheise replied:
I can see the Java code is being used now, but doesn't that mean that the mixin can be deleted?

defsub replied:
The mixin pairs with the corresponding custom actions. If they are being used, the mixin can be used.  When using the original actions (non-custom) for `stop`, `fast forward`, and `rewind`, on either Android 13+ or prior, the mixin isn't needed.  The java code makes them work on Android 13+ and at the same keeps things working on older Android versions.

ryanheise replied:
I'll have to look at it again, but I think I was suspicious of whether this was stale code since the action codes for these actions were not being used anywhere in either the plugin or the examples.

defsub replied:
I can see that. I think what's confusing is that from the Dart code either native or custom controls can now be created. And to keep some backward compatibility, deprecated native controls created on the Dart side are handled as custom controls in Java.

ryanheise replied:
I've had another look at it, and I'm fairly sure this logic should not be duplicated in two places. Actually when I looked at it this time I noticed that the logic had been duplicated in 3 places (once in native, and twice in Dart).

I'll branch and make some changes to your code to refine this aspect of it and then share it with you for feedback.

ryanheise commented:
I've made some further changes to your PR in this branch `feature/custom_actions_android13`:

- Backwards compatible actions are now handled in one place (native side)
- Renamed CustomAction to CustomMediaAction (consistent with Dart-side naming, and avoiding native-side clash)
- Renamed actions to controls in AudioService wherever they referred to Dart-side controls.

I think this should work, because if an app needs to customise the behaviour of `stop`, `fastForward` or `rewind` in Android 13, they can still do it in the same way that worked for Android 12, and they can still have the same control over the order in which controls are presented.

Let me know if I've missed anything, and of course let me know if it still works on your device, both the backwards compatible aspect and the `favorite` custom action.

Note, I also deleted the custom actions example since the `favorite` example was already included in the android 13 example.

defsub commented:
Thanks @ryanheise. I did a quick review and it looks good. I will do some testing tomorrow as well.

In other news, not sure if you're aware that your audio_service and just_audio work great on Wear OS. Another topic but happy to discuss w/ you later.  Thanks.

ryanheise commented:
WearOS probably depends on the vendor a bit, I've tested it on a Samsung watch and it was a bit flakey getting it to recognise the media session at first but once it did connect, it worked. Do you have a watch recommendation that works well?

defsub commented:
I'm good with your changes. I tested on Android 9 (Samsung J3) and 13 (Pixel 6 Pro) as well. 

I have a Samsung Watch 5 LTE (SM-R915U) 44mm with Android 11 API 30. The only issues I've had are related to the network. Streaming over bluetooth via the phone can be slow. However using wifi or mobile is great. Local files and remote work fine.

hovermouse commented:
Hi - I wondered if this PR was going to be merged, I would love to be able to implement custom actions on the notification and this looks like it will fix that?
Thanks!

ryanheise commented:
Yes, it is the next PR in line to be merged. Please give the issue another bump if it is not merged by the end of this weekend.

dimakrest commented:
@ryanheise @defsub Thanks guys for an awesome job and for improving this great plugin! I am waiting for this one too, since I have to bump my target sdk to 33 by end of this month :)

ryanheise commented:
I have just created a copy of your branch in this repo called `defsub-custom_actions_android13` where I fixed up the merge conflicts and fixed some lint warnings.

@defsub can you remind me again why we are implementing the custom action dispatch in Java and not on the Dart side?

```java
        @Override
        public void onCustomAction(String action, Bundle extras) {
            if (listener == null) return;
            if (CUSTOM_ACTION_STOP.equals(action)) {
                listener.onStop();
            } else if (CUSTOM_ACTION_FAST_FORWARD.equals(action)) {
                listener.onFastForward();
            } else if (CUSTOM_ACTION_REWIND.equals(action)) {
                listener.onRewind();
            } else {
                listener.onCustomAction(action, extras);
            }
        }
```

Also, there doesn't seem to be much difference between `example_android13.dart` and `example_custom_action.dart` - can we get away with one example or is there a reason to have two?

defsub commented:
@ryanheise I recall you had [feature/custom_actions_android13] where you made been incorporating changes - see back on May 29 in this thread. In that branch you had mentioned you deleted the custom actions example. Also, the reason for the dispatch in Java is for backward compatibility.

ryanheise commented:
Thanks, it's been a while, but that rings a bell. I will check when I get home.

ryanheise commented:
I've reapplied those changes over to the `feature/custom_actions_android13` branch.

I was slightly concerned that the additional `custom` value in the enum would technically be a breaking change, but I think it'll be fine. The issue with adding a value to an enum is that any app that does a case-complete switch will then get told by the compiler that their switch is no longer complete. Although as it turns out, we happen to have a private enum value `_setShuffleModeEnabled` which would prevent complete switches anyway. Fortunate :-)

Anyway, please let me know if the latest commit on `feature/custom_actions_android13` works or whether anyone has any feedback. Note that until I publish the new platform interface to pub, you will need to manually comment/uncomment the local development versions of the dependencies in the plugin's `pubspec.yaml` files in order to run it (see the comments inside `pubspec.yaml`).

hovermouse commented:
Hi @ryanheise - I've tested (on emulator and physical device - on both pre-Android 13 and Android 13 devices) and it's looking good from my side.
Thanks for pushing this and thanks for a great plugin!

hovermouse commented:
Hi @ryanheise and @defsub - I know this isn't the place for coding support but I'm going slightly nuts trying to call a CustomAction from my home screen widget.  I just can't get my head around the Android side which I have never worked in.

I am using the [home_widget](https://pub.dev/packages/home_widget) package but not really using any features other than to get the widget deployed. My widget layout echos the Notification UI with a bit more info added in.

With this update, in my player notification, I now have REW - PLAY/PAUSE - FF plus a customAction BOOKMARK.  
And the functions work great - even showing on Android Auto.

In my `AppWidgetProvider()` of my home widget code, I figured out that I can get a PendingIntent by calling 
`MediaButtonReceiver.buildMediaButtonPendingIntent(context, PlaybackStateCompat.ACTION_REWIND)`
(the home_widget package sets things up in Kotlin)

This call just seems to be a helper to create a `PendingIntent.getBroadcast()` call and I can follow the logic that it hits the `MediaButtonReceiver` in the `AudioService` code by pulling the package from context and looking for receivers that handle **ACTION_MEDIA_BUTTON**.

I can add that intent onto the onclick of an icon in my widget like this
`setOnClickPendingIntent(R.id.widget_rewind_button, intent)`
And again, all this works brilliantly to control the player.

What I can't figure, and I've been at this for 3 days now, is how to call a customAction on AudioService, and I think that's mainly because I can't get my head around 'intents' and 'actions'. 
Looking through your code, I see the builder for a `MediaButtonPendingIntent` which follows the pattern I'm using for the main buttons, but I can't see a comparable for `customAction`. 

I've been all around the docs and the closest I have got is this
```
private fun getBookmarkIntent(context: Context): PendingIntent? {
    val customActionIntent = Intent(context, com.ryanheise.audioservice.AudioService::class.java)
    customActionIntent.action = "bookmark"
    val pendingIntentCustomAction = PendingIntent.getService(context, 0, customActionIntent, PendingIntent.FLAG_IMMUTABLE)
    return pendingIntentCustomAction
       }
```
...but this clearly isn't a customAction (and it doesn't work!). 
I'd really appreciate a steer on this - and again, sorry for polluting the PR chat with this.

ryanheise commented:
audio_service is intended to be used entirely from Dart/Flutter. Presumably home_widget lets you handle the button press in Dart code, so your Dart handler for the button press should be the place to write code to interact with audio_service, and then you will be able to refer to the example for how the Dart code would work.

hovermouse commented:
Thanks for the response, the home_widget plugin creates its own receiver which in turn created it's own issues for me - ie if I click play the app starts up but not using AudioService and so not initialised as I need it to be.  
Using the MediaButton events fixes that (for PLAY/FF/REW) because AudioService is the receiver.  
I've fudged it for now so I can move on!

defsub commented:
@ryanheise I tested the example_android13 on a Pixel 6 pro running Android 13 and Samsung J3 running Android 9. Works great on both. It's important again to note that custom actions only work on Android 13 and higher. I think it's possible to make them work prior to Android 13 by targeting lower SDK versions, however, it won't work with Android 13 w/ the same build. I did try to target lower than 13 but ran into dependency issues. My opinion is that since custom actions are new, it's fair to say they only work on Android 13 or higher. I'm good with the current PR.

ryanheise commented:
Thanks for testing that. I'm just doing some final checks before release. The platform interface version is set to 0.2.0 but based on the discussion above I'll set that to 0.1.1 to avoid breaking any dependency constraints with `pub`.

ryanheise commented:
Merged and released as 0.18.11. Thanks @defsub for the original PR and working with me on the backward/future compatibility issues. Also to those who helped to test.

Note that in future Android version bumps, I plan to separate API design changes or new features from the version bump itself where possible.
